Fix render pipeline lighting issue with small cubes

sekwah41 opened the pull request:
This will also allow other mods which need to calculate small normalised vectors to be able to use the default classes without needing to write their own, while not affecting any side effects created by how the base game uses this function.

![image](https://user-images.githubusercontent.com/3209834/137609776-a4dd39a8-ede6-49dd-847a-3d083df76d28.png)

It fixes the bug shown above and is useful for other mods where any problems with the resulting vector are probably not as bad as it not being calculated in the first place.


After changes
![image](https://user-images.githubusercontent.com/3209834/137609942-5b59b589-de5c-4817-bd6a-24b288b228b6.png)

sekwah41 committed:
Fix render pipeline lighting issue with small cubes

This will also allow other mods which need to calculate small normalised vectors to be able to use the default classes without needing to write their own, while not effecting any side effects created by how the base game uses this function.

marchermans commented:
So I know that this fixes it, since this is how I fixed this in C&B.
But would this not also require a Mojang Minecraft Issue ticket?

sekwah41 commented:
I'm don't know if it creates a problem in vanilla. So far I've only found it only causes a problem in the experimental forge render pipeline.

I'm not sure how quick or if mojang would fix it unless a vanilla problem is found.

gigaherz commented:
If the bug is in vanilla code we request that the issues are reported to mojang. Even if it's unlikely they will fix them, we want them to be reported.

We can then mark forge's fix as `// Forge: fix MC-xxxxx`

sekwah41 commented:
> If the bug is in vanilla code we request that the issues are reported to mojang. Even if it's unlikely they will fix them, we want them to be reported.
> 
> We can then mark forge's fix as `// Forge: fix MC-xxxxx`

I'll report it when I'm back later then. Though I'll take a quick look at where this function is used in vanilla to see if I can find anything it might break.

sekwah41 commented:
I've created the issue https://bugs.mojang.com/browse/MC-239212

I will also change the code to just avoid it when the length is 0 rather than avoiding it entirely, seeing as normalising doesn't work in that case anyway.

sekwah41 committed:
Change to not entirely disable checks

Allows to still check for a 0 length vector. After checking what its used for, nowhere uses the return boolean it seems and its its only ever large vectors or rendering.

gigaherz commented:
I'm wondering if it would be worth doing instead, `if ((double)f < Float.MIN_NORMAL)` which would mean the value is either denormalized or zero.

pupnewfster commented:
> I'm wondering if it would be worth doing instead, `if ((double)f < Float.MIN_NORMAL)` which would mean the value is either denormalized or zero.

At that point shouldn't it be `if (f < Float.MIN_NORMAL)` so as to not potentially have issues due to float to double floating point precision? At least given the fact it seems to use it as a float still lower down so wouldn't have that precision issue show up there instead.

pupnewfster commented on patches/minecraft/com/mojang/math/Vector3f.java.patch:
```diff
@@ -1,16 +1,11 @@
        float f = this.f_122228_ * this.f_122228_ + this.f_122229_ * this.f_122229_ + this.f_122230_ * this.f_122230_;
 -      if ((double)f < 1.0E-5D) {
-+      if (!force && (double)f < 1.0E-5D) {
++      if ((double)f == 0) {
```
At the end of this line you should include a comment like: `//Forge: Fix MC-239212`

sekwah41 commented:
> I'm wondering if it would be worth doing instead, `if ((double)f < Float.MIN_NORMAL)` which would mean the value is either denormalized or zero.

Forgot that existed, that would probably be the best

sekwah41 committed:
Add reference to mc issue and switch to Float.MIN_NORMAL

sekwah41 commented:
After taking a deeper look into the code, the only time Mojang ever calls it, they actually perform another function on it which doesn't seem to matter if it's normalised or not.

Other than that they normalise some vectors which are pre-defined.

I hope they fix it but with that in mind, altering this and making it part of forge should not cause any problems in mc and should only benefit any mods using these classes rather than implementing their own.
